### PR TITLE
lib:simulations:Merge_sort :Merge Sort implemented #174

### PR DESCRIPTION
**Pull request by mrmayurgithub**
@yashugarg I've added Merge algorithm simulation for solving the issue #174  , it works fine.

### CODE REVIEW

**Comment by mrmayurgithub** [comment]
@yashugarg Don't merge the PR right now, I'm making some changes.

**Comment by mrmayurgithub** [comment]
@yashugarg please merge this PR.

File: lib/src/data/simulations.dart [hunk 1]
```diff
@@ -13,22 +14,23 @@ import 'package:simulate/src/simulations/epicycloid_curve.dart';
...
-    6: "epicycloid curve pattern mathematics animation pencil lines ",
-    7: "epicycloid curve pattern mathematics animation ",
-    8: "maurer rose pattern mathematics animation",
+    3: "merge sort algorithm sorting bars ",
+    4: "rose pattern mathematics sequence ",
+    5: "fourier series mathematics ",
+    6: "lissajous curve pattern mathematics animation ",
+    7: "epicycloid curve pattern mathematics animation pencil lines ",
+    8: "epicycloid curve pattern mathematics animation ",
+    9: "maurer rose pattern mathematics animation",
```
Comment (yashlamba) [inline lib/src/data/simulations.dart]: Don't add merge sort in between, add it to the end of the list.

File: lib/src/data/simulations.dart [hunk 2]
```diff
@@ -54,92 +54,82 @@ class Simulations with ChangeNotifier {
...
         fav: _favorites[2],
       ),
       SimulationCard(
         id: 3,
+        simulationName: 'Merge Sort',
+        image: theme.darkTheme ? 'assets/simulations/MergeSortDark.png' : 'assets/simulations/MergeSortLight.png',
+        direct: MergeSort(),
+        infoLink: 'https://en.wikipedia.org/wiki/Insertion_sort',
+        fav: _favorites[3],
+      ),
```
Comment (yashlamba) [inline lib/src/data/simulations.dart]: here too, add merge sort to the end :)

File: lib/src/simulations/merge_sort.dart [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+                  ),
+                ),
+                // Positioned(
+                //   top: 5,
+                //   left: 5,
+                //   child: Text(
+                //     "Comparisons: $counter \nMax: ${_elements[i]} \nArray Iteration: ${_elements.length - n + 1}",
+                //     style: Theme.of(context).textTheme.subtitle2,
+                //   ),
+                // ),
```
Comment (yashlamba) [inline lib/src/simulations/merge_sort.dart]: Why is this commented?

File: lib/src/simulations/merge_sort.dart [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+              onPressed: () {
+                pause = !pause;
+                //TODO: Fix the bug, i.e, when we pause the code shows unusual behaviour
+                if (!sort)
+                  _sort();
+                else
+                  setState(() {
+                    sort = false;
+                  });
+              },
```
Comment (yashlamba) [inline lib/src/simulations/merge_sort.dart:218]: The pause bug needs to be fixed, when paused, it restarts the sort.

File: lib/src/simulations/merge_sort.dart [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        await Future.delayed(Duration(milliseconds: delay2));
+
+        containers[i] = Container(
+          color: Colors.greenAccent[400],
+          height: (elements[i] + 0.5),
+          width: barwidth,
+        );
+        streamContainer.add(containers);
+      }
+  }
```
Comment (yashlamba) [inline lib/src/simulations/merge_sort.dart:140]: I don't think the sorted array is turning green at the end.

**Review by yashlamba — CHANGES_REQUESTED** [review]
Really good work, loved the simulation! Still needs some work, I have added a few points. Thanks for your efforts!

File: lib/src/simulations/merge_sort.dart (hunk 3 above)
Comment (mrmayurgithub) [reply lib/src/simulations/merge_sort.dart]: Oh sorry, I forgot to make the changes.

File: lib/src/simulations/merge_sort.dart (hunk 4 above)
Comment (mrmayurgithub) [reply lib/src/simulations/merge_sort.dart:218]: Ok

File: lib/src/simulations/merge_sort.dart (hunk 5 above)
Comment (mrmayurgithub) [reply lib/src/simulations/merge_sort.dart:140]: I didn't understand ? In the simulation of bubble sort, we're making the array green at the end.

File: lib/src/simulations/merge_sort.dart (hunk 4 above)
Comment (mrmayurgithub) [reply lib/src/simulations/merge_sort.dart:218]: Yes, it actually scans the array from the starting and I tried making some other changes but I wasn't able to implement the pause function properly. 
Hopefully, someone else will be able to resolve this issue. :)

**Pull request merged**